openmc/tallies.py

```diff
@@ -1110,7 +1110,11 @@ def get_filter_indices(self, filters=[], filter_bins=[]):
                     bins = self_filter.bins
 
             # Add indices for each bin in this Filter to the list
-            indices = np.array([self_filter.get_bin_index(b) for b in bins])
+
+            if type(self_filter) in filters:
+                indices = np.array([self_filter.get_bin_index(b) for b in bins])
+            else:
+                indices = np.arange(len(bins))
```

Comment:
I'm wondering if we can restructure the entire body of the main for loop here as:
```Python
for i, self_filter in enumerate(self.filters):
    for j, test_filter in enumerate(filters):
        if type(self_filter) is test_filter:
            bins = filter_bins[j]
            indices = np.array([self_filter.get_bin_index(b) for b in bins])
            break
    else:
        indices = np.arange(self_filter.num_bins)
```
Would that be equivalent to this?

Reply:
Nearly! I like how it cleans up the loop. The only difference in behavior I can see would be for the `EnergyFunctionFilter` where we could potentially return an array with length > 1 if the requested bins has length > 1 for some reason. So if we want to preserve that behavior we may want:

```python
for i, self_filter in enumerate(self.filters):
    for j, test_filter in enumerate(filters):
        # always use default case for EnergyFunctionFilter
        if type(self_filter, openmc.EnergyFunctionFilter):
            continue
        if type(self_filter) is test_filter:
            bins = filter_bins[j] 
            indices = np.array([self_filter.get_bin_index(b) for b in bins]) 
            break 
    else: 
        indices = np.arange(self_filter.num_bins)

```

Given that the filter doesn't have any bins, I think it's pretty rare that it would be passed in the `filters` argument though.